File: sort-tool/go_sort_tool/sorter/algorithms/radix_sort.go
package algorithms

func RadixSort(arr []string) {
    if len(arr) <= 1 {
        return
    }

    // Find the maximum length string
    maxLen := len(arr[0])
    for _, s := range arr {
        if len(s) > maxLen {
            maxLen = len(s)
        }
    }

    // Perform counting sort for every character from right to left
    for i := maxLen - 1; i >= 0; i-- {
        countingSort(arr,i)
    }
}

func countingSort(arr []string, position int) {
    n := len(arr)
    output := make([]string, n)
    count := make([]int, 256) // assuming ascii chars

    // Store count of occurrences
    for _, s := range arr {
        index := 0
        if position < len(s) {
            index = int(s[position])
        }
        count[index]++
    }

    // Change count[i] so that count[i] now contains actual
    // position of this character in output array
    for i := 1; i < 256; i++ {
        count[i] += count[i-1]
    }

    // Build the output array
    for i := n - 1; i >= 0; i-- {
        index := 0
        if position < len(arr[i]) {
            index = int(arr[i][position])
        }
        output[count[index]-1] = arr[i]
        count[index]--
    }

    // Copy the output array to arr[], so that arr[] now
    // contains sorted characters
    for i := 0; i < n; i++ {
        arr[i] = output[i]
    }
}

File: sort-tool/rust_sort_tool/src/algorithms/radix_sort.rs
pub fn sort(arr: &mut [u32]) {
    if arr.is_empty() {
        return;
    }
    
    let max = *arr.iter().max().unwrap();
    let mut exp = 1;

    while max / exp > 0 {
        counting_sort(arr, exp);
        exp *= 10;
    }
}

fn counting_sort(arr: &mut [u32], exp: u32) {
    let mut output = vec![0; arr.len()];
    let mut count = vec![0; 10];

    // Store count of occurrences in count[]
    for &num in arr.iter() {
        count[((num / exp) % 10) as usize] += 1;
    }

    // Change count[i] so that count[i] now contains
    // actual position of this digit in output[]
    for i in 1..10 {
        count[i] += count[i - 1];
    }

    // build the output array
    for &num in arr.iter().rev() {
        let index = ((num / exp) % 10) as usize;
        output[count[index] - 1] = num;
        count[index] -= 1;
    }

    // copy the output array to arr[], so that arr[] now
    // contains sorted numbers according to current digit
    arr.copy_from_slice(&output);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_radix_sort_empty() {
        let mut arr: Vec<u32> = vec![];
        sort(&mut arr);
        assert_eq!(arr, vec![]);
    }

    #[test]
    fn test_radix_sort_single_element() {
        let mut arr = vec![1];
        sort(&mut arr);
        assert_eq!(arr, vec![1]);
    }

    #[test]
    fn test_radix_sort_sorted_array() {
        let mut arr = vec![1, 2, 3, 4, 5];
        sort(&mut arr);
        assert_eq!(arr, vec![1, 2, 3, 4, 5]);
    }

    #[test]
    fn test_radix_sort_reverse_sorted_array() {
        let mut arr = vec![5, 4, 3, 2, 1];
        sort(&mut arr);
        assert_eq!(arr, vec![1, 2, 3, 4, 5]);
    }

    #[test] 
    fn test_radix_sort_unsorted_array() {
        let mut arr = vec![170, 45, 75, 90, 802, 24, 2, 66];
        sort(&mut arr);
        assert_eq!(arr, vec![2, 24, 45, 66, 75, 90, 170, 802]);
    }

    #[test]
    fn test_radix_sort_with_zeros() {
        let mut arr = vec![0, 10, 1, 100, 1000, 10000];
        sort(&mut arr);
        assert_eq!(arr, vec![0, 1, 10, 100, 1000, 10000]);
    }

    #[test]
    fn test_radix_sort_large_numbers() {
        let mut arr = vec![1_000_000, 10_000, 100_000, 1_000, 10_000_000];
        sort(&mut arr);
        assert_eq!(arr, vec![1_000, 10_000, 100_000, 1_000_000, 10_000_000]);
    }
}
